RFC: sorting specification

trinity-1686a opened the pull request:
### Description

This PR tries to formalize the state of sorting/top-k in Quickwit, and propose changes both in behavior and in code organization to enable a more correct, simpler to maintain and simpler to optimize behavior.
The goal for now is to gather opinions on the state of sorting, and how it should move forward. Feel free to comment if you disagree with the proposed new behavior

### How was this PR tested?

For now, this is a text-only PR. It was prompted by tests done [here](https://github.com/quickwit-oss/quickwit/pull/3545#issuecomment-1662380517)

saroh commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+The special value `_score` means sorting by score, with a sort that is then Descending by default
+(documents matching better goes first).
+
+In case of equality between two documents, the GlobalDocId, composed of (SplitId, SegmentId, DocId)
```
Hey :wave:,
This might be based on wrong assumptions, disregard if so :D
With [cursor pagination](https://betterprogramming.pub/understanding-the-offset-and-cursor-pagination-8ddc54d10d98?gi=72175bea8ea5) in mind: shouldn't the `timestamp` be somewhere in the tie breaker, espacially if it's the posting's sort order? It probably is as it seems like an optim for top-k on your prefered "timestamp" sort?

trinity-1686a replied:
I think if you want `timestamp` to be used as a tie breaker of sorts, you should put it last in the `sort_by` query parameter. Right now it's limited to two fields, but that limitation will be lifted off at some point. I don't think it would help in optimization to have `timestamp` as a tie breaker, on the contrary it means you have to always download that fastfield. However a "more stable sort order" is something that would be nice to have

saroh commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+
+If we allow unsorted requests, we can go further and stop searching as soon as we have k hits,
+without even looking at other splits metadata. Argument can be made in favor of this because
+GlobalDocId is not stable, and can change during a merge, so order is not guaranteed anyway,
```
> GlobalDocId is not stable
:'( 
Good to know.

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+## Behavior
+
+Sorting is controlled by the `sort_by` query parameter. It accepts a comma separated list of fields
+to use for sorting. Sorting is Ascending by default. The sorting order can be reversed by prefixing
```
Crazy idea:

Should we want descending by default?

In elasticsearch it is descending by default for _score, ascending for everything else.
For timestamp, descending is better too, so maybe descending by default is a good idea.

fulmicoton replied:
Aaaaaaah I had missed this was a description of the current state!

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+Sorting is controlled by the `sort_by` query parameter. It accepts a comma separated list of fields
+to use for sorting. Sorting is Ascending by default. The sorting order can be reversed by prefixing
+a field name with a hyphen `-`.
+The special value `_score` means sorting by score, with a sort that is then Descending by default
```
ah yes you spotted that.

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+(documents matching better goes first).
+
+In case of equality between two documents, the GlobalDocId, composed of (SplitId, SegmentId, DocId)
+is used as a tie breaker. It is always used in Ascending order *inside a split*. It is used in the
```
This behavior is not great. You are offering a spec where the repartition of docs in splits as an effect on search results.

Could we do ascending always?

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+
+If a document doesn't have a value for a sorting field, that document is considered to go after any
+document which has a value, independantly of sort order. That is, when sorting the value 1,2 and
+None, ascending sort would give `[1, 2, None]`, and descending sort would give `[2, 1, None]`.
```
yes

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+document which has a value, independantly of sort order. That is, when sorting the value 1,2 and
+None, ascending sort would give `[1, 2, None]`, and descending sort would give `[2, 1, None]`.
+
+If the client does not request sorting, documents are sorted using (DocId, SplitId, SegmentId), in
```
`(SplitId, SegmentId, DocId)` would make way more sense no?

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+document which has a value, independantly of sort order. That is, when sorting the value 1,2 and
+None, ascending sort would give `[1, 2, None]`, and descending sort would give `[2, 1, None]`.
+
+If the client does not request sorting, documents are sorted using (DocId, SplitId, SegmentId), in
+Descending order.
```
I suspect you made an error.

```suggestion
If the client does not request sorting, documents are sorted using (SplitId, SegmentId, DocId), in
Descending order. In other words, everything happens as if documents were sorted by a constant value.
```

(This suggests that descending order by default is not a bad idea)

trinity-1686a replied:
as surprising as it seems, this is the actual order used. We use `(SplitId, SegmentId, DocId)` as a tie breaker, but sort primarily based on a [value which is the DocId](https://github.com/quickwit-oss/quickwit/blob/5d2c66e38e17ba5952196f01123d87f5a98b6a3c/quickwit/quickwit-search/src/collector.rs#L203). It might be more explicit to write that we sort by `(DocId, SplitId, SegmentId, DocId)`, which is a bit redundant, but more of what's actually happening

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+is used as a tie breaker. It is used to sort in the same order as the first field being sorted by.
+This means it is in Ascending order by default, except when the first field being sorted by is
+`_score` .
+// TODO or should it always be Ascending by default? I don't think it matters much, but would require
```
I suspect descending by default would be better.

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+some more special handling of `_score`.
+
+If a document doesn't have a value for a sorting field, that document is considered to go after any
+document which has a value, independantly of sort order. That is, when sorting the value 1,2 and
```
```suggestion
document which has a value, independently of sort order. That is, when sorting the value 1,2 and
```

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+document which has a value, independantly of sort order. That is, when sorting the value 1,2 and
+None, ascending sort would give `[1, 2, None]`, and descending sort would give `[2, 1, None]`.
+
+If a client does not request sorting, documents are sorted using (SplitId, SegmentId, DocId), on
```
I think it is easier to define that behavior by saying "everything should happen as we sorted by a constant field"
in descending order. (Assuming we go for descending by default.)

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+(documents matching better goes first).
+
+In case of equality between two documents, the GlobalDocId, composed of (SplitId, SegmentId, DocId)
+is used as a tie breaker. It is used to sort in the same order as the first field being sorted by.
```
ok

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,93 @@
+If we allow unsorted requests, we can go further and stop searching as soon as we have k hits,
+without even looking at other splits metadata. Argument can be made in favor of this because
+GlobalDocId is not stable, and can change during a merge, so order is not guaranteed anyway,
+at least not until Quickwit has support for a Point In Time mechanism.
```
Now, when paginating using the scroll API we have such a PIT mechanism.

trinity-1686a replied:
I've seen that, but it's not really usable from outside the scroll API, is it? I meant mostly for a "general purpose PIT" like the one ES has, and recommend using with things like search-after

fulmicoton commented:
Great job! I suspect we might prefer Descending order for everything by default but otherwise I did not have any comment.

trinity-1686a commented:
I've added some test for the sort logic inside a split, as well as fixed the test for sort merging (the ascending test wasn't performing an asc sort, and the test vector were ordered in a way that made it hard to notice any issue). I also fixed a bug previously mentioned (though I otherwise kept the old sorting behaviour).
If nobody has an objection to the new ~~world~~ sort order, I think this is ready to merge, with a followup implementing the changes described here

fulmicoton commented on docs/internals/sorting.md:
```diff
@@ -0,0 +1,84 @@
+It also tries to describe optimizations that may be enabled (but are not necessarily implemente)
+by this behavior.
+
+# Current state
```
I think we can remove the `Current State`
